### ISSUE TITLE
Newbie error: Cloud init not working?

### ISSUE THREAD

**Issue by dankegel** [lxc/lxd#5787]
* Distribution: ubuntu
 * Distribution version: 18.04 or 19.04
   * LXC version: 3.0.3
   * LXD version: 3.0.3

# Issue description

Newbie found shiny toy cloud-config and tried to use it, but failed.

Inspiration was https://discuss.linuxcontainers.org/t/cloud-init-pre-configure-not-working-on-lxd-3-0-bionic-beaver/1765/3

1) Create the script touchy.sh containing:

```
#!/bin/sh
UID=$(id -u)
GID=$(id -g)
lxc profile create touchy || true
cat <<_EOF_ | lxc profile edit touchy
config:
  raw.idmap: both $UID $GID
  user.user-data: |
    #cloud-config
    runcmd:
      - [touch, /tmp/kilroy-was-here]
description: Touchy LXD profile
_EOF_

lxc launch --profile default --profile touchy ubuntu:19.04 touchy-ubu1904
# Have to sleep, or we get the error
# su: user ubuntu does not exist
#sleep 5
lxc exec touchy-ubu1904 -- su ubuntu -c "ls /tmp"

```

2) sh touchy.sh

3) Observe the error
```su: user ubuntu does not exist```

4) Curse, uncomment the sleep line, stop and remove touchy-ubu1904, and run the script again.

5) Observe the error
```ls: cannot access '/tmp/kilroy-was-here': No such file or directory```

6) Verify that cloud-init is in fact preinstalled inside the container (it is) and
that the init data makes it to /var/lib/cloud/seed/nocloud-net/user-data (it does).

7) scratch head, wonder if there's a typo in the script

**Comment by stgraber** [comment]
Yeah, it's a bit confusing but it can all be explained I think :)

The user indeed isn't created until cloud-init is done running. LXD has no visibility on when that is, `lxc start` will complete as soon as the init process was spawned. The sleep is one way around that, the other would be a loop doing something like `systemctl -a | grep cloud-init` to wait until cloud-init is actually done.

The `/tmp` issue can very likely be explained by the fact that most Linux distributions have a systemd unit cleaning up `/tmp` at boot time. Assuming that this unit runs after cloud-init would explain why your file never seems to make it. Try creating it in `/root` and hopefully things will work fine then.

**Comment by dankegel** [comment]
Thanks for the handholding :-)

This worked on my little i3-4130 with ubuntu 18.04, though it took 22 seconds for kilroy to finally show up:
```
#!/bin/sh
UID=$(id -u)
GID=$(id -g)
lxc profile create touchy || true
cat <<_EOF_ | lxc profile edit touchy
config:
  raw.idmap: both $UID $GID
  user.user-data: |
    #cloud-config
    runcmd:
      - [touch, /home/ubuntu/kilroy-was-here]
description: Touchy LXD profile
_EOF_

lxc stop touchy-ubu1904
lxc delete touchy-ubu1904
lxc launch --profile default --profile touchy ubuntu:19.04 touchy-ubu1904
n=0
while ! lxc exec touchy-ubu1904 -- su ubuntu -c "ls -l /home/ubuntu/kilroy-was-here" | grep -v grep | grep kilroy
do
  sleep 1
  n=$((n + 1))
done
echo "It took $n seconds for kilroy to show up."
````

Are the arguments to raw.idmap correct? I didn't want to hardcode them, and the script failed
on an i9-9900K with ubuntu 18.04, complaining

```
buildbot@foo $ lxc info --show-log local:touchy-ubu1904
...
lxc touchy-ubu1904 20190522032713.215 ERROR    conf - conf.c:lxc_map_ids:2999 - newuidmap failed to write mapping "newuidmap: uid range [1000-1001) -> [1000-1001) not allowed": newuidmap 31106 0 165536 1000 1000 1000 1 1001 166537 64535
...
buildbot@foo $ cat /etc/subuid
buildbot:100000:65536
lxd:165536:65536
root:165536:65536
```
As you noted the subuid stuff is hard for people to understand, and I r people.

**Comment by stgraber** [comment]
`raw.idmap` is `both UID-IN-HOST UID-IN-CONTAINER`

In your case, I think you actually want:
`raw.idmap` set to:
```
uid $(id -u) 1000
gid $(id -g) 1000
```

The long delay isn't unusual as cloud-init will run things like `apt update` on first boot, which hits the network and can easily take 15-20s to complete.

**Comment by dankegel** [comment]
Thanks.  I tried
```
#!/bin/sh
UID=$(id -u)
GID=$(id -g)
lxc profile create touchy || true
cat <<_EOF_ | lxc profile edit touchy
config:
  raw.idmap: "uid $UID 1000\ngid $GID 1000"
  user.user-data: |
    #cloud-config
    runcmd:
      - [touch, /home/ubuntu/kilroy-was-here]
description: Touchy LXD profile
_EOF_

lxc stop touchy-ubu1904
lxc delete touchy-ubu1904
lxc launch --profile default --profile touchy ubuntu:19.04 touchy-ubu1904
n=0
while ! lxc exec touchy-ubu1904 -- su ubuntu -c "ls -l /home/ubuntu/kilroy-was-here" | grep -v grep | grep kilroy
do
  sleep 1
  n=$((n + 1))
done
echo "It took $n seconds for kilroy to show up."
```
but it still failed with "newuidmap: uid range [1000-1001) -> [1000-1001) not allowed".

I tried switching to snap-installed lxd, but after that, even ```lxc ls``` fails with
```
Error: Get http://unix.socket/1.0: dial unix /var/lib/lxd/unix.socket: connect: connection refused
```
even after rebooting, and making sure 'id' shows I'm in the lxd group.  After reading https://stgraber.org/2016/10/17/lxd-snap-available/ I purgedapt lxd and lxc-client,
and now "lxc ls" works... as does the script.

So, as you suggested, installing lxd via snap seems to avoid that check.  I wonder what I was
doing wrong.

Case closed, I think, given that ubuntu 19.04 has ditched non-snap lxd.  On to the next toe-stubbing :-)

But if anyone has a principled recipe for avoiding that subuid problem, please post it, thanks.
